---
title: "test_out_archR"
author: "T Attenborough"
date: "2023-10-04"
output: html_document
---


```{r}
library(Signac)
library(Seurat)
library(bambu)
library(MACSr)
library(tidyverse)
library(ArchR)
```

```{r}
addArchRThreads(threads = 4)
```

```{r}
inputFiles <- getTutorialData("Hematopoiesis")
inputFiles
```

```{r}
addArchRGenome("hg19")
```

Creating Arrow Files
Now we will create our Arrow files which will take 10-15 minutes. For each sample, this step will:

Read accessible fragments from the provided input files.
Calculate quality control information for each cell (i.e. TSS enrichment scores and nucleosome info).
Filter cells based on quality control parameters.
Create a genome-wide TileMatrix using 500-bp bins.
Create a GeneScoreMatrix using the custom geneAnnotation that was defined when we called addArchRGenome().

```{r}
ArrowFiles <- createArrowFiles(
  inputFiles = inputFiles,
  sampleNames = names(inputFiles),
  filterTSS = 4, #Dont set this too high because you can always increase later
  filterFrags = 1000, 
  addTileMat = TRUE,
  addGeneScoreMat = TRUE
)
```

We can inspect the ArrowFiles object to see that it is actually just a character vector of Arrow file paths.
```{r}
ArrowFiles
```

Strict quality control (QC) of scATAC-seq data is essential to remove the contribution of low-quality cells. In ArchR, we consider three characteristics of data:

The number of unique nuclear fragments (i.e. not mapping to mitochondrial DNA).
The signal-to-background ratio. Low signal-to-background ratio is often attributed to dead or dying cells which have de-chromatinzed DNA which allows for random transposition genome-wide.
The fragment size distribution. Due to nucleosomal periodicity, we expect to see depletion of fragments that are the length of DNA wrapped around a nucleosome (approximately 147 bp).
For our tutorial data, an example of these QC plots is shown below for the BMMC sample:

Inferring Doublets
After Arrow file creation, we can infer potential doublets (a single droplet containing multiple cells) that can confound downstream results. This is done using the addDoubletScores() function.
```{r}
doubScores <- addDoubletScores(
  input = ArrowFiles,
  k = 10, #Refers to how many cells near a "pseudo-doublet" to count.
  knnMethod = "UMAP", #Refers to the embedding to use for nearest neighbor search.
  LSIMethod = 1
)
```

Creating an ArchRProject 
With our Arrow files in hand, we are now ready to create an ArchRProject. An ArchRProject is associated with a set of Arrow files and is the backbone of nearly all ArchR analyses.

```{r}
proj <- ArchRProject(
  ArrowFiles = ArrowFiles, 
  outputDirectory = "HemeTutorial",
  copyArrows = TRUE #This is recommened so that you maintain an unaltered copy for later usage.
)
```

We can also ask which data matrices are available within the ArchRProject which will be useful downstream once we start adding to this project:

```{r}
getAvailableMatrices(proj)
```

Now we can filter putative doublets based on the previously determined doublet scores using the filterDoublets() function. This doesn’t physically remove data from the Arrow files but rather tells the ArchRProject to ignore these cells for downstream analysis.
```{r}
proj <- filterDoublets(ArchRProj = proj)
```

Dimensionality Reduction and Clustering
ArchR implements an iterative LSI dimensionality reduction via the addIterativeLSI() function.
```{r}
proj <- addIterativeLSI(ArchRProj = proj, useMatrix = "TileMatrix", name = "IterativeLSI")
```


To call clusters in this reduced dimension sub-space, we use the addClusters() function which uses Seurat’s graph clustering as the default clustering method.
```{r}
proj <- addClusters(input = proj, reducedDims = "IterativeLSI")
```

Visualizing in a 2D UMAP Embedding
We can visualize our scATAC-seq data using a 2-dimensional representation such as Uniform Manifold Approximation and Projection (UMAP). To do this, we add a UMAP embedding to our ArchRProject object with the addUMAP() function. This function uses the uwot package to perform UMAP.
```{r}
proj <- addUMAP(ArchRProj = proj, reducedDims = "IterativeLSI")
```

Using this UMAP, we can visualize various attributes of our cells which are stored in a matrix called cellColData in our ArchRProject. To do this, we use the plotEmbedding() function and we specify the variable to use for coloration via a combination of the colorBy and name parameters.

For example, we can color by “Sample”:
```{r}
p1 <- plotEmbedding(ArchRProj = proj, colorBy = "cellColData", name = "Sample", embedding = "UMAP")
p1
```

```{r}
p2 <- plotEmbedding(ArchRProj = proj, colorBy = "cellColData", name = "Clusters", embedding = "UMAP")
p2
```

```{r}
ggAlignPlots(p1, p2, type = "h")
```

To save an editable vectorized version of this plot, we use the plotPDF() function.
```{r}
plotPDF(p1,p2, name = "Plot-UMAP-Sample-Clusters.pdf",
        ArchRProj = proj, addDOC = FALSE, width = 5, height = 5)
```

Assigning Clusters with Gene Scores
We can try to assign biological labels to these clusters using marker genes of known hematopoietic regulators. First, we add imputation weights using MAGIC to help smooth the dropout noise in our gene scores.
```{r}
proj <- addImputeWeights(proj)
```

Now we can overlay our marker gene scores on our 2D UMAP embedding.
```{r}
markerGenes  <- c(
    "CD34",  #Early Progenitor
    "GATA1", #Erythroid
    "PAX5", "MS4A1", "MME", #B-Cell Trajectory
    "CD14", "MPO", #Monocytes
    "CD3D", "CD8A"#TCells
  )

p <- plotEmbedding(
    ArchRProj = proj, 
    colorBy = "GeneScoreMatrix", 
    name = markerGenes, 
    embedding = "UMAP",
    imputeWeights = getImputeWeights(proj)
)

```

```{r}
p$CD14
```

To plot all genes we can use cowplot to arrange the 9 different plots together. Each of these marker genes lights up the corresponding cell clusters. For example, we infer that the cells that have the highest gene score for CD3D, a known T cell marker, are in fact T cells.

```{r}
#Rearrange for grid plotting
p2 <- lapply(p, function(x){
    x + guides(color = FALSE, fill = FALSE) + 
    theme_ArchR(baseSize = 6.5) +
    theme(plot.margin = unit(c(0, 0, 0, 0), "cm")) +
    theme(
        axis.text.x=element_blank(), 
        axis.ticks.x=element_blank(), 
        axis.text.y=element_blank(), 
        axis.ticks.y=element_blank()
    )
})
do.call(cowplot::plot_grid, c(list(ncol = 3),p2))
```

To save an editable vectorized version of this plot, we use the plotPDF() function.
```{r}
plotPDF(plotList = p, 
    name = "Plot-UMAP-Marker-Genes-W-myImputation.pdf", 
    ArchRProj = proj, 
    addDOC = FALSE, width = 5, height = 5)
```

Visualizing Genome Browser Tracks
In addition to plotting gene scores per cell as a UMAP overlay, we can browse the local chromatin accessibility at these marker genes on a per cluster basis with genome browser tracks. To do this, we use the plotBrowserTrack() function which will create a list of plots, one for each of the genes specified by markerGenes.
```{r}
p <- plotBrowserTrack(
    ArchRProj = proj, 
    groupBy = "Clusters", 
    geneSymbol = markerGenes, 
    upstream = 50000,
    downstream = 50000
)
```

To plot a track of a specific gene, we can simply select one from the list.
```{r}
grid::grid.newpage()
grid::grid.draw(p$CD14)
```

We can save a multi-page PDF with a single page for each gene locus in our plot list using the plotPDF() function
```{r}
plotPDF(plotList = p, 
    name = "Plot-Tracks-Marker-myGenes.pdf", 
    ArchRProj = proj, 
    addDOC = FALSE, width = 5, height = 5)
```

Last but certainly not least, ArchR natively supports an interactive and dynamic genome browser that can be launched locally via a shiny app. To do this, we use the ArchRBrowser() function.
```{r}
ArchRBrowser(ArchRProj = proj)
```

Saving and Loading an ArchRProject 
To easily save an ArchRProject for later use or for sharing with collaborators, we use the saveArchRProject() function. This copies the current ArchRProject object and all of the Arrow files to a specified directory. If we don’t specify an output directory (as below), saveArchRProject() uses the output directory that we specified upon creation of our ArchRProject. In this case that is the folder “HemeTutorial”.
```{r}
proj <- saveArchRProject(ArchRProj = proj)
```

When we are ready to load this saved ArchRProject we use the loadArchRProject() object and provide the path to the folder containing the saved ArchRProject object.
```{r}
proj <- loadArchRProject(path = "HemeTutorial")
```


